node/src/chain/fsm.rs

```diff
@@ -779,30 +778,28 @@ impl Debug for BlockRequest {
     }
 }
 
-/// Requests a block by height from a specified peer
-async fn request_block<N: Network>(
+/// Requests a block by height/hash from the network with so-called
+/// Flood-request approach.
+async fn flood_request_block<N: Network>(
     network: &Arc<RwLock<N>>,
     req: BlockRequest,
-    peer_addr: SocketAddr,
 ) {
-    let mut get_data = GetData::default();
+    // Request only one resource/block
+    let mut inv = Inv::new(1);
     match req {
         BlockRequest::ByHeight(height) => {
-            get_data.inner.add_block_from_height(height);
+            inv.add_block_from_height(height);
         }
         BlockRequest::ByHash(hash) => {
-            get_data.inner.add_block_from_hash(hash);
+            // Request from network the full block, if it is missing then
+            // request the candidate block.
```

Comment:
I think we should leave to the requester the choice of what resource to ask for: if the Quorum is known, it should ask for the Candidate only.

Reply:
If it asks only for the candidate, it may happen that the receiver node does have the full block only but not the candidate.

Additionally, if the sender sends the full block then the certificate cache will not be touched at all.

Reply:
> If it asks only for the candidate, it may happen that the receiver node does have the full block only but not the candidate.

If it has the full block, it implicitly has the candidate too. It's just the block without the certificate.

I think Candidate and Attestation (we should update the nomenclature for this) should be decoupled. Especially considering the fact that the actual Certificate (i.e., the "chosen" Attestation) is included in the following block. 
In fact, keeping the Attestation in the block is unnecessary (we're basically storing two equivalent objects).
But this is to be addressed in a separate DIP/issue.

Given the current implementation, I guess we could either return the block without the certificate, or the full block (but we're kind of wasting bandwidth).

> Additionally, if the sender sends the full block then the certificate cache will not be touched at all.

I'm not sure I get the point here... I guess the cache is not affected because we would already have the full block. But this would be equally true if we received the Candidate only (bc we would have both candidate and attestation).

Reply:
> If it has the full block, it implicitly has the candidate too. It's just the block without the certificate.

Valid point! The issue #1713 is actually suggesting the same.

> Given the current implementation, I guess we could either return the block without the certificate, or the full block (but we're kind of wasting bandwidth).

Sending a block message with/without certificate will not impact significantly the bandwidth as the certificate size is only  ~100 bytes. 

> I'm not sure I get the point here... I guess the cache is not affected because we would already have the full block. But this would be equally true if we received the Candidate only (bc we would have both candidate and attestation).

By  cache I meant the HashMap (it'll be replaced with RocksDB CF) storing the received certificates. If we receive the Candidate only, we rely on the cache for having the certificate.

Another side note: If we ask only for the Candidate that means we fully trust the `unverified` Quorum msg. 

I'll reconsider both options.

Reply:
In the case "we received a Quorum and we don't know the candidate", I don't see much difference in asking a full block or asking the Candidate, in terms of trust. In both cases we're kind of trusting the Quorum (which is why we should do a sanity check).

Similarly, for the cache, it would be the same to check the received Candidate against the stored certificate or against the one attached to the full block. In both cases we end up with the block accepted (or tried to) and the corresponding certificate out of the cache.
This is assuming everything is correct, of course.

A possible advantage of always sending the full block, if available, could be to have a valid certificate in case the one we received is wrong. Although I'm not sure there's any meaningful threat here...
I'll take the chance to do some thinking on possible "Fake Quorum" attacks

Reply:
>  I don't see much difference in asking a full block or asking the Candidate, in terms of trust

If an attacker sends a Quorum msg with a correct `candidate_hash` but invalid certificate (sort of a Fake Quorum) then it will disallow the acceptance of the block for the particular peer. That would be resolved if Quorum msg verification is performed but that's not the case yet.

Additionally, due to redundancy factor (K_BETA=3) we can expect to have at least one honest peer of 3 that will respond with valid (full) block.

To  avoid wasting time on premature optimizations, let me recap:

We started with your suggestion:
```
I think we should leave to the requester the choice of what resource to ask for: if the Quorum is known, it should ask for the Candidate only.
```

- As mentioned above, the difference is around 100 bytes between full block and candidate block - a minor part of the bandwidth.
- We should consider this only after having the verification of quorum msg incorporated.

Reply:
I'm fine with that. Let's make nodes respond with the full block if available and consider changes when sanity check on quorums is implemented.